sycl/source/detail/usm/opencl_shim.cpp

```diff
@@ -17,16 +17,32 @@ using namespace cl::sycl::detail::usm;
 
  **/
 
-static clHostMemAllocINTEL_fn pfn_clHostMemAllocINTEL = NULL;
+// Changing to per-context tracking
+// TODO: use piContext everywhere
+static std::map<cl_context, clHostMemAllocINTEL_fn> pfn_clHostMemAllocINTEL;
```

Comment:
Please avoid using global objects. It seems you can easily store them in context object or in some object which depends on context object.

Reply:
Sure.

Reply:
On second thought.  By the time we're here, we've left sycl land.  We don't have a handle on the sycl context.  Not sure I can avoid global objects here unless we break the CL extension interface in a really non-CL friendly way.

Reply:
Let's take hostMemAlloc for example, we can have something like this:

```
// The class holds all data you currently using globals for.
// CLUSM and USMHandle can be merged.
class USMHandle  {
  FuncType *pfn_clDeviceMemAllocINTEL = nullptr;
  ...
  CLUSM CLUSM;
};
class context_impl {

  context_impl () {
    ...
    MUSMHandle.reset(getUSMHandleFor(m_Context));
  }
void *alignedAlloc(...) {
  // The following logic can be implemented in USMHandle
  if(m_USMHandle->pfn_clDeviceMemAllocINTEL) {
    pfn_clHostMemAllocINTEL(m_Context, ...);
  }
  else {
    USMHandle->CLUSM->hostMemAlloc(...);
  }
}
...
unique_ptr<USMHandle> m_USMHandle;
};
```

Reply:
Yeah, this is what I meant.  Now we're driving all this branching logic into the implementation instead of having a nice clean shim.  Would you guys prefer this?  It's a non-trivial rewrite.

Reply:
Probably the only way forward here is to also do the PI rewrite - this would let us do a clean interface using SYCL objects instead of CL objects and let us do this type of book keeping.

Reply:
I would prefer this as having global objects cause a lot of problems.